src/osd/PrimaryLogPG.cc

```diff
@@ -8523,9 +8523,21 @@ void PrimaryLogPG::_do_rollback_to(OpContext *ctx, ObjectContextRef rollback_to,
   map<snapid_t, interval_set<uint64_t> >::iterator iter =
     snapset.clone_overlap.lower_bound(snapid);
   ceph_assert(iter != snapset.clone_overlap.end());
-  interval_set<uint64_t> overlaps = iter->second;
+
+  // adjust overlap
+  auto last_clone_overlap_iter = --snapset.clone_overlap.end();
+  interval_set<uint64_t> rollback_overlaps;
+  rollback_overlaps.insert(0, rollback_to->obs.oi.size);
+  last_clone_overlap_iter->second.swap(rollback_overlaps);
+
+  // adjust size
+  auto last_clone_size_iter = --snapset.clone_size.end();
+  last_clone_size_iter->second = rollback_to->obs.oi.size;
```

Comment:
I'm not sure ```last_clone_size_iter->second = rollback_to->obs.oi.size;``` is correct.  last_clone_size_iter should be the size of the most recent clone before HEAD -- rollback (or any other mutation) shouldn't change that.

The next statement down ```overlaps.insert(0, rollback_to->obs.oi.size);```, however, should probably be something like ```overlaps.insert(0, MIN(rollback_to->obs.oi.size, last_clone_overlap_iter->second));```, so perhaps it just seems to work because it fixed the overlaps initialization?